### PR TITLE
Add @polka/compression package

### PR DESCRIPTION
**Pull request by developit**
This implements gzip + brotli (native) compression as a new `@polka/compression` package. It should be ready to go, with tests and docs.

### CODE REVIEW

**Comment by lukeed** [comment]
Legend, thank you sir!

I'll pass through tomorrow for the failing Node 8.x suite and do some docs shuffling. Looks great though 🙌

**Comment by developit** [comment]
Ah so the failing suite is because on Node < 12.7, this package automatically disables brotli (since its unsupported). I'll have the test check.

**Comment by developit** [comment]
Can probably get coverage back up by doing a test that uses `createReadStream(file).pipe(res)`.

**Comment by developit** [comment]
I broke CI somehow - tests should be passing on 8/10/12 though (see 0f6bcbd).

**Comment by newtriks** [comment]
After reading through open PR's I can see you're most likely very busy, just wanted to see if there was a chance at some point to merge @developit's work @lukeed?

**Comment by developit** [comment]
IIRC Luke found and fixed a bug in the version of this that we have inlined in WMR, I need to backport that here.

**Comment by patak-dev** [comment]
Sorry for bumping this @lukeed @developit, but FYI Vite is also going to inline this PR (once we merge https://github.com/vitejs/vite/pull/6557). If there is a chance to get it merged at one point, we could reduce the maintenance burden for both WMR and Vite.

**Comment by Kikobeats** [comment]
what a good PR 👏 @lukeed could be nice if you can take a look and move this forward

**Comment by pke** [comment]
While I like the implementation I wonder if such should be part of an application server or better be left to reverse proxies like nginx. For me applications servers only handle the application and not logging, tracing, auditing, caching or compression.

**Comment by developit** [comment]
@pke There are numerous cases where compression is necessary in an application server. Nobody should have to run a reverse proxy just for local development (as Patak alluded to above).

**Comment by pke** [comment]
I never needed compression during local development, what would be some of the use cases? I try to understand.

Beside that I am all for merging this, but the project seems to be pretty much dead and unmaintained. I might switch back to koa, although its not express middleware compatible.

**Comment by lukeed** [comment]
Stable and dead are different things. There are two flavors of Polka and both work well for different use cases. 

This PR still needs me to backport some fixes I/we found in the wmr integration & tests to solidify those fixes long term.

**Comment by pke** [comment]
Nice to see you here @lukeed ;) I just thought a 2 year old PR, which is green is not merged seemed like a dead project. But I didn't understand the implications of this PRs merge.

There are also a bunch of issues open without any response (incl one of me ;) so if there is anything we could help with to get "next" out of beta I'd be glad to assist.

**Comment by Kikobeats** [comment]
The implementation code of this PR is really appreciated, there is no real alternatives these days for taking advantage of brotli, so I decided to just put it in standalone package and anyone can take advantage of it

https://github.com/Kikobeats/http-compression

Full credits to @developit for writing that simple and elegant code, I promise to maintain the code and keep it evolving

File: packages/compression/index.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+				// backpressure
+				compress.on('data', (...args) => write.apply(res, args) === false && compress.pause());
+				on.call(res, 'drain', () => compress.resume());
+				compress.on('end', (...args) => end.apply(res, args));
+			}
+
+			const listeners = pendingListeners;
+			pendingListeners = null;
+			listeners.forEach(p => on.apply(res, p));
+			if (pendingStatus) writeHead.call(res, pendingStatus);
```
Comment (benmccann) [inline packages/compression/index.js:68]: I see that @lukeed updated this line in `wmr`: https://github.com/preactjs/wmr/commit/23bee83c34aaea472a3329d3ed09b4e14c0d5dd8

Comment (benmccann) [reply packages/compression/index.js:68]: It's also been updated in Vite's version: https://github.com/vitejs/vite/blob/eef9da13d0028161eacc0ea699988814f29a56e4/packages/vite/src/node/server/middlewares/compression.ts#L80

**Comment by benmccann** [comment]
This version already contains the main fix that Vite copied from wmr (https://github.com/preactjs/wmr/pull/155) when they did their initial commit. It is missing one other fix though that I pointed out above (https://github.com/lukeed/polka/pull/148#discussion_r1515297995).

Once that change makes its way in, I think it'd be great to merge and release this! It looks like the version that Vite copied into their codebase almost exactly two years ago has been running there without any fixes necessary in the time since, so this is probably pretty stable. It also looks like this would fix the issue we were having in SvelteKit with the `compression` library, so I would love to be able to use this!

**Comment by Kikobeats** [comment]
@benmccann any particular reason you don't want to use [Kikobeats/http-compression](https://github.com/Kikobeats/http-compression)? Same source code, shipped as npm package, running for years.

**Comment by pke** [comment]
@Kikobeats thanks for putting this in a module, which seems good way to keep polka small and lean. Then this PR can be closed @developit?

**Review by lukeed — APPROVED** [review]
Thank you @benmccann for the team ping to resurrect this (and multiple others for trying over the lifetime of this PR!)

Long ago, when this PR initially appeared, I had some quirks that I (and rest of `wmr` team) were trying to chase down. I've had a few iterations of this running in production over the years – but wasn't great at organizing my findings on all of them – and wanted to ensure confidence in this (v important) package before releasing. Great to see that the `wmr` version (which wasn't this PR) has been proven stable by Vite/WMR/others' usage all this time.

Of course, a massive thank-you to @developit who was brave enough to tackle this problem in the first place & (patiently) entrusted the work over to the Polka household. Thank you 🐐 🦊 🏆

**Comment by benmccann** [comment]
Amazing!! Thank you so much @lukeed for getting this across the finish line and making it part of `@polka` and @developit for contributing it!

**Pull request merged**